### PR TITLE
GAP 2079: Update emails in Find

### PR DESCRIPTION
**Pull request by kiramarstonTCO**
## Description

Ticket # and link: https://technologyprogramme.atlassian.net/jira/software/projects/GAP/boards/216?selectedIssue=GAP-2079

Summary of the changes and the related issue. List any dependencies that are required for this change: Added new design for notifying the user of an email mismatch between the user-service DB and OneLogin. Changes the user-service DB entry to match OneLogin and then redirects the user appropriately.

## Type of change

Please check the relevant options.

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] This change requires a documentation update.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes:

- [x] Unit Test

- [ ] Integration Test (if applicable)

- [ ] End to End Test (if applicable)

## Screenshots (if appropriate):

Please attach screenshots of the change if it is a UI change:

https://github.com/cabinetoffice/gap-user-service/assets/107405237/89a0b861-348b-4059-abad-e1f0433d5889

# Checklist:

- [ ] If I have listed dependencies above, I have ensured that they are present in the target branch.
- [x] I have performed a self-review of my code.
- [x] I have commented my code in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation where applicable.
- [ ] I have ran cypress tests and they all pass locally.

### CODE REVIEW

File: src/main/java/gov/cabinetofice/gapuserservice/enums/LoginJourneyState.java [hunk 1]
```diff
@@ -1,20 +1,16 @@
 package gov.cabinetofice.gapuserservice.enums;
 
 import gov.cabinetofice.gapuserservice.model.RoleEnum;
-import gov.cabinetofice.gapuserservice.model.User;
-import gov.cabinetofice.gapuserservice.service.OneLoginService;
-import org.slf4j.Logger;
 
+// TODO inject logger here?
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/enums/LoginJourneyState.java:5]: Can you remove the TODO comment please

File: .github/workflows/feature.yml [hunk 2]
```diff
@@ -5,15 +5,13 @@ on:
     branches:
       - feature/**
       - bug/**
-      - bugfix/**
```
Comment (dylanwrightCO) [inline .github/workflows/feature.yml:8]: Can we add these back in?

File: src/main/java/gov/cabinetofice/gapuserservice/enums/LoginJourneyState.java [hunk 3]
```diff
@@ -101,10 +85,18 @@ public LoginJourneyRedirect getLoginJourneyRedirect(final RoleEnum role) {
...
-                                           final String jwt,
-                                           final Logger logger) {
+        public LoginJourneyState nextState(final NextStateArgs nextStateArgs) {
+            if (nextStateArgs.userInfo() != null) {
+                final String newEmail = nextStateArgs.userInfo().getEmailAddress();
+                final String oldEmail = nextStateArgs.user().getEmailAddress();
+                final boolean emailsDiffer = !newEmail.equals(oldEmail);
+
+                if (emailsDiffer) {
+                    nextStateArgs.oneLoginService().setUsersEmail(nextStateArgs.user(), newEmail);
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/enums/LoginJourneyState.java:95]: There is no unit test coverage for this

File: src/main/resources/application.properties [hunk 4]
```diff
@@ -55,7 +55,6 @@ onelogin.service-redirect-url=redirectUrlValue
 onelogin.private-key=privateKeyValue
 feature.onelogin.enabled=false
 feature.onelogin.migration.enabled=false
-feature.validate-user-roles-in-middleware=true
```
Comment (dylanwrightCO) [inline src/main/resources/application.properties:58]: Why are we removing this?

File: .github/workflows/feature.yml (hunk 2 above)
Comment (kiramarstonTCO) [reply .github/workflows/feature.yml:8]: This came from one of my develop merges, not sure what it's about myself...

File: src/main/resources/application.properties (hunk 4 above)
Comment (kiramarstonTCO) [reply src/main/resources/application.properties:58]: Another develop merge oddity, I'm looking into it

File: src/main/java/gov/cabinetofice/gapuserservice/enums/LoginJourneyState.java (hunk 1 above)
Comment (kiramarstonTCO) [reply src/main/java/gov/cabinetofice/gapuserservice/enums/LoginJourneyState.java]: Done!

File: .github/workflows/feature.yml (hunk 2 above)
Comment (kiramarstonTCO) [reply .github/workflows/feature.yml]: Reverted :)

File: src/main/resources/application.properties (hunk 4 above)
Comment (kiramarstonTCO) [reply src/main/resources/application.properties]: Reverted :)

File: src/main/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2.java [hunk 5]
```diff
@@ -247,10 +260,16 @@ private void verifyStateAndNonce(final String nonce, final StateCookieDto stateC
...
+    private String runStateMachine(final String redirectUrlCookie, final User user, final String jwt,
+            final boolean hasAcceptedPrivacyPolicy, final OneLoginUserInfoDto userInfo) {
+        log.debug(loggingUtils.getLogMessage("Running state machine", 5), redirectUrlCookie, user, jwt,
+                hasAcceptedPrivacyPolicy, userInfo);
+
+        String redirectUrl = user.getLoginJourneyState()
+                .nextState(new NextStateArgs(oneLoginService, user, jwt, log, hasAcceptedPrivacyPolicy, userInfo))
                 .getLoginJourneyRedirect(user.getHighestRole().getName())
                 .getRedirectUrl(adminBaseUrl, applicantBaseUrl, techSupportAppBaseUrl, redirectUrlCookie);
+        log.debug(loggingUtils.getLogMessage("Redirecting to: ", 1), redirectUrl);
```
Comment (dominicwest) [inline src/main/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2.java:272]: Love these new logs 👌

File: src/main/java/gov/cabinetofice/gapuserservice/enums/LoginJourneyState.java (hunk 3 above)
Comment (kiramarstonTCO) [reply src/main/java/gov/cabinetofice/gapuserservice/enums/LoginJourneyState.java:94]: There is now! 👍

File: src/main/java/gov/cabinetofice/gapuserservice/enums/LoginJourneyState.java [hunk 6]
```diff
@@ -117,9 +108,22 @@ public LoginJourneyRedirect getLoginJourneyRedirect(final RoleEnum role) {
                 case APPLICANT, FIND -> LoginJourneyRedirect.APPLICANT_APP;
             };
         }
+    },
+
+    MIGRATING_FIND_EMAILS {
+        @Override
+        public LoginJourneyState nextState(final NextStateArgs nextStateArgs) {
+            nextStateArgs.oneLoginService().setUsersLoginJourneyState(nextStateArgs.user(), USER_READY);
```
Comment (dominicwest) [inline src/main/java/gov/cabinetofice/gapuserservice/enums/LoginJourneyState.java:116]: Assuming the migration of finds users is to be done in a separate ticket or PR?

**Review by dominicwest — APPROVED** [review]

**Review by dylanwrightCO — APPROVED** [review]

**Pull request merged**